SAM E70: fix Ethernet driver cache coherency issues and always enable data cache

aurel32 opened the pull request:
This pull request fixes the Atmel SAM Ethernet driver cache coherency issues by using the recently added nocache memory and always enable the data cache. After this the `gmac_desc_*` functions introduced to manage the cache coherency through cache operations look a bit empty. I wonder if we should get rid of them. The same way given the cache is now always enabled, I wonder if we should still check if it is enabled. Anyway this can probably be done in a further cleanup PR.

Fixes #9812

tgorochowik reviewed (changes requested):
Nice! That looks and works just great!

Yeah, the helpers are not really needed that much now, and checking if cache is enabled may be redundant, but I guess it is ok to clean it up later.

Could we at least remove this comment from the header?
https://github.com/zephyrproject-rtos/zephyr/blob/fa5b8b4b506f056c86be9a9a216852dbbbaa66c4/drivers/ethernet/eth_sam_gmac.c#L19

aurel32 commented:
@tgorochowik Thanks for the review and thanks for pointing at that comment. I have just pushed a new version which removes it.

tgorochowik reviewed (approved):
Thanks!

mnkp commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -152,27 +150,25 @@ static inline void dcache_clean(u32_t addr, u32_t size)
 {
 	desc->w0 = value;
-	dcache_clean((u32_t)desc, sizeof(struct gmac_desc));
+	__DSB();
```
I believe this line is redundant. In a few places in the driver code where we do care about the order of the memory accesses there is the DMB instruction. So DSB is no longer necessary. Unless there is another issue at hand I'm not aware of.

jukkar reviewed (approved):
Tested this and it works ok. Also tried with a version that removed `__DSB()` calls as suggested by @mnkp , and did not see any issues there either.

jukkar commented:
Hmm, just after giving my comment, I started to see gptp sync issues. I am using sam-e70 board for gptp testing. I will do some more testing with this patch and a version that removes the __DSB() calls.

jukkar commented:
The gptp drops from sync eventually regardless of this patch or not, so the issues I am seeing are related to some other problem. +1 from me to this fix.

aurel32 commented:
@mnkp indeed, it looks like the necessary barriers are there, so i'll drop the __DSB() from the next version. Now the `gmac_desc_*` functions are really empty...

I also have found another issue looking at the code, `dcache_clean` is called from a second place in addition to the net buffers: 

https://github.com/zephyrproject-rtos/zephyr/blob/724d6f79092e8a2639563f38e97e71f1389eb39b/drivers/ethernet/eth_sam_gmac.c#L1170-L1174

This is done on a 32-bit value, so less than a cache line, and therefore looks worrying. This has been added in commit a313e5c74f0 by @tgorochowik . I don't really get why as from what I understand the  RX buffer accounting list is not a list shared with Ethernet device. I guess we have two choices there, either drop that cache operation or declare `rx_frag_list_que0/1/2` as nocache.

aurel32 commented:
I ran some additional tests, and it some cases the Ethernet device get stuck for some time, or indefinitely, probably something similar to #11255.

I have found it happens when the Zephyr network stack has to sent 2 packets in row, for example an ack for the just received data + answer data. This seems to be a latent issue, but it happens more often given that instruction cache and now data cache have been enabled. Basically the CPU is now faster than Ethernet device (which is good in some sense).

The problem seems to be the following one:
1. the first packet goes through `eth_tx`, one `tx_desc_sem` semaphore is taken, and the `tx_timeout_work` delayed work is submitted.
2. the second packet also goes through `eth_tx`, another `tx_desc_sem` semaphore is taken, and the `tx_timeout_work` delayed work is not started because there is already one already submitted. Not sure that's a problem in that case.
3. the first packet has been sent, one `tx_desc_sem` semaphore is given and the `tx_timeout_work` delayed work is cancelled.
4. the second packet has been sent but given the delayed work has already been cancelled, `tx_completed` is not called: the `tx_desc_sem` semaphore is not given back and the network packet is not unreferenced.

I am not fully sure about how exactly the steps 2 and 3 are interleaved, what is sure is that the call to `tx_completed` is skipped for the second packet has the job has already been cancelled. Adding some delay between sending packets (for instance at the beginning of `eth_tx`) workarounds the issue.

It looks like we should not allow sending a new packet until the previous packet has been fully sent.

mnkp commented:
@aurel32 Good catch. `dcache_clean` on line 1172 is indeed not needed. Just as you mentioned "the RX buffer accounting list is not a list shared with Ethernet device".

I'll have a look at the other bug you described.

aurel32 commented:
@mnkp, thanks for the confirmation, i'll add a patch removing it.

For the bug, here's the workaround I am using:

```
diff --git a/drivers/ethernet/eth_sam_gmac.c b/drivers/ethernet/eth_sam_gmac.c
index e0e2d4a883..751dfc66ef 100644
--- a/drivers/ethernet/eth_sam_gmac.c
+++ b/drivers/ethernet/eth_sam_gmac.c
@@ -1329,6 +1327,11 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
         */
        net_buf_push(pkt->frags, net_pkt_ll_reserve(pkt));
 
+       /* Wait until the previous packet has been fully sent */
+       while (k_sem_count_get(&queue->tx_desc_sem) != (queue->tx_desc_list.len - 1)) {
+               k_yield();
+       }

```

aurel32 commented:
Note, I am able to reproduce the issue using `samples/net/sockets/dumb_http_server` and running a loop of 100 wget fetching data from it.

aurel32 committed:
drivers: eth: gmac: drop wrong cache clean operation

The RX buffer accounting list is not a list shared with Ethernet device,
therefore there is no need to run a cache clean operation on it.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
drivers: eth: gmac: use non-cached memory for descriptor lists

Atmel SAM Ethernet module is using a scatter-gather technique to
exchange data with the Ethernet driver. To communicate the location
of the data buffers the driver sets-up a so called descriptor list. This
is effectively a place in RAM containing a sequence of 32-bit words
representing buffer location and its status.

Currently the cache coherency is handled using cache clean or
invalidate. Unfortunately this strategy only works correctly when the
corresponding data size is a multiple of the cache line. This is not
the case here and might lead to data loss or corruption.

Instead of using cache operations, this patch moves the descriptor
listed to the recently added non-cached memory region, as recommended by
ATMEL. A data synchronisation barrier is still required for writes, as
the non-cached memory is defined with TEX=1, i.e. it is not strongly
ordered. The descriptor lists alignment can be decreased to 4 bytes, as
required by the SAM Ethernet module.

The RX/TX buffer are left unchanged, still managed by cache operations.

Fixes #9812

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
drivers: eth: gmac: get rid of descriptors helpers

Now that the descriptor lists are in non-cached memory, the helpers are
just accessing the w0 or w1 members of the structure. Just drop them and
access the members directly.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

mnkp commented:
> 2\. the second packet also goes through `eth_tx`, another `tx_desc_sem` semaphore is taken, and the `tx_timeout_work` delayed work is not started because there is already one already submitted. Not sure that's a problem in that case.

Well spotted. The fact that the `k_delayed_work_submit()` is not called for the second frame is indeed the problem. And as you notice

> 4\. the second packet has been sent but given the delayed work has already been cancelled, `tx_completed` is not called:

The whole timeout concept has to be reworked. We shouldn't wait until the tx frame is fully sent. In any case the bug is orthogonal to this PR.

mnkp commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -515,8 +467,7 @@ static void tx_completed(Gmac *gmac, struct gmac_queue *queue)
 
-	__ASSERT(gmac_desc_get_w1(&tx_desc_list->buf[tx_desc_list->tail])
-		 & GMAC_TXW1_USED,
+	__ASSERT(&tx_desc_list->buf[tx_desc_list->tail]->w1 & GMAC_TXW1_USED,
```
`tx_desc_list->buf[tx_desc_list->tail].w1`

aurel32 replied:
Good catch, that's fixed now.

aurel32 commented:
> The whole timeout concept has to be reworked. We shouldn't wait until the tx frame is fully sent. In any case the bug is orthogonal to this PR.

Not fully orthogonal. Enabling the cache means the code sending packets runs faster, triggering the  issue. Maybe I should drop the last commit enabling the cache from the PR?

mnkp commented:
> Not fully orthogonal.

True. I guess one of the maintainers should decide if we should fix the timeout bug first or merge partially this PR.

As for the timeout bug - I'm not sure if we should discuss it further here or maybe move the conversation to a separate issue - one possible and easy fix would be to get rid of all the k_delayed_* functions and associated data structures and inside eth_tx where we have
```
		k_sem_take(&queue->tx_desc_sem, K_FOREVER);
```
replace `K_FOREVER` with the timeout value. It works a bit differently than the old design but does the job. We can only get stuck on the tx_queue if there is a bug in the driver or the hardware malfunctions. So this is a safety feature, not part of the normal functional behavior. We would only have to be careful writing the small piece of code that cleans up the data in case we do fail to get the semaphore on time.

aurel32 commented:
This PR is ready, but should go after the fix for #11255, hence the DNM label.

aurel32 commented:
PR #11255 has now been fixed, but this PR now also depends on PR #12548. The DNM label should therefore stay.

aurel32 commented:
This issue has been ready and approved for almost two months, but is blocked by another PR. As the freeze is almost there, I believe it make sense to merge it anyway. I have therefore removed the last commit enabling the cache, I'll do a separate PR for it that can be merged once PR #12817 gets in.
